libsql-storage/src/lib.rs

```diff
@@ -201,8 +196,16 @@ impl Wal for DurableWal {
         // - save the current max_frame_no for this txn
         trace!("DurableWal::begin_read_txn()");
         let rt = tokio::runtime::Handle::current();
-        let frame_no = tokio::task::block_in_place(|| rt.block_on(self.frames_count()));
-        self.max_frame_no = frame_no;
+        let remote_frame_no = tokio::task::block_in_place(|| rt.block_on(self.frames_count()));
+        let local_frame_no = self.local_cache.get_max_frame_num().unwrap();
+        if local_frame_no < remote_frame_no {
+            warn!(
+                "cache (max_frame_num = {:?}) is lagging behind the remote(max_frame_num = {:?})!",
+                local_frame_no, remote_frame_no
+            )
+            // TODO: replicate data from source
+        }
+        self.max_frame_no = local_frame_no;
```

Comment:
In the follow up PR, I will switch to `remote` mode if I find primary is lagging.